cases/src/org/commcare/cases/instance/CaseDataInstance.java

```diff
@@ -0,0 +1,83 @@
+package org.commcare.cases.instance;
+
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.model.instance.utils.FormLoadingUtils;
+import org.javarosa.core.services.Logger;
+import org.javarosa.xml.util.InvalidStructureException;
+
+import java.io.IOException;
+
+/**
+ * An external data instance that respects CaseDB template specifications.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class CaseDataInstance extends ExternalDataInstance {
+    private static TreeElement caseDbSpecTemplate = null;
+    private static final String CASEDB_WILD_CARD = "CASEDB_WILD_CARD";
+    private static final Object lock = new Object();
+
+    public CaseDataInstance() {
+        // For serialization
+    }
+
+    /**
+     * Copy constructor
+     */
+    public CaseDataInstance(ExternalDataInstance instance) {
+        super(instance);
+    }
+
+    /**
+     * Does the reference follow the statically defined CaseDB spec?
+     */
+    public boolean hasTemplatePath(TreeReference ref) {
+        synchronized (lock) {
```

Comment:
Is it overkill to synchronize the whole template path check method? We need to synchronize both reads & writes, not just writes, correct?